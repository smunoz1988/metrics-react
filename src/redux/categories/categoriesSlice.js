import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  loading: false,
  categories: [],
  error: '',
};

export const getCategories = createAsyncThunk('category/getCategories', async () => {
  try {
    const config = {
      headers: {
        'X-RapidAPI-Key': '595740a7f5mshc92914bbcf4727fp18322djsnaf6c8af4b471',
        'X-RapidAPI-Host': 'fia-formula-1-championship-statistics.p.rapidapi.com',
      },
    };

    const url = 'https://fia-formula-1-championship-statistics.p.rapidapi.com/api/drivers/';

    const resp = await axios.get(url, config);
    return resp.data;
  } catch (error) {
    return error;
  }
});

const categoriesSlice = createSlice({
  name: 'category',
  initialState,
  reducers: {

  },
  extraReducers: (builder) => {
    builder.addCase(getCategories.pending, (state) => {
      state.loading = true;
    });
    builder.addCase(getCategories.fulfilled, (state, action) => {
      state.loading = false;
      state.categories = action.payload;
      state.error = '';
    });
    builder.addCase(getCategories.rejected, (state, action) => {
      state.loading = false;
      state.categories = [];
      state.error = action.error.message;
    });
  },
});

export default categoriesSlice.reducer;
