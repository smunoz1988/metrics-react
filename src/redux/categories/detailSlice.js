import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  loading: false,
  details: [],
  error: '',
};

export const getDetails = createAsyncThunk('category/getCategories', async (driver) => {
  try {
    const config = {
      headers: {
        'X-RapidAPI-Key': '4743a0dcfcmsh6774803b4a66ca6p1f5f3ejsn67ea319599fd',
        'X-RapidAPI-Host': 'fia-formula-1-championship-statistics.p.rapidapi.com',
      },
    };

    const url = `https://fia-formula-1-championship-statistics.p.rapidapi.com/api/drivers/detaills${driver}`;

    const resp = await axios.get(url, config);
    return resp.data;
  } catch (error) {
    return error;
  }
});

const detailsSlice = createSlice({
  name: 'detail',
  initialState,
  reducers: {

  },
  extraReducers: (builder) => {
    builder.addCase(getDetails.pending, (state) => {
      state.loading = true;
    });
    builder.addCase(getDetails.fulfilled, (state, action) => {
      state.loading = false;
      state.details = action.payload;
      state.error = '';
    });
    builder.addCase(getDetails.rejected, (state, action) => {
      state.loading = false;
      state.details = [];
      state.error = action.error.message;
    });
  },
});

export default detailsSlice.reducer;