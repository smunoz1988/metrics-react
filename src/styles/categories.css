.cardsContainer {
  display: grid;
  grid-template-columns: auto auto;
}

.driverContainer {
  border: solid black 0.1rem;
  text-align: center;
}

input {
  width: 80%;
  height: 2rem;
  margin: auto;
  margin-bottom: 0.5rem;
  border: solid 0.05rem black;
  padding-left: 0.5rem;
}