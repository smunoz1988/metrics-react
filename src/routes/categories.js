import { useSelector, useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { TbHelmet } from 'react-icons/tb';
import { IoIosArrowBack, IoMdMic } from 'react-icons/io';
import { AiTwotoneSetting } from 'react-icons/ai';
import { getCategories } from '../redux/categories/categoriesSlice';
import '../styles/categories.css';

const Categories = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const driversData = useSelector((store) => store.category?.categories?.drivers);
  const [filterDrivers, setFilterDrivers] = useState('');

  useEffect(() => {
    dispatch(getCategories());
  }, [dispatch]);

  const handleChange = (event) => {
    const { value } = event.target;
    setFilterDrivers(value);
  };

  const handleClickItem = (driverId) => {
    navigate(`/Details/${driverId}`);
  };

  if (!driversData) {
    return <div>Loading...</div>;
  }

  const filteredDrivers = driversData.filter((driver) => {
    const lowercaseFirstName = driver.firstname.toLowerCase();
    const lowercaseLastName = driver.lastname.toLowerCase();
    const lowercaseFilter = filterDrivers.toLowerCase();

    return lowercaseFirstName.includes(lowercaseFilter) || lowercaseLastName.includes(lowercaseFilter);
  });

  return (
    <>
      <div className="header">
        <IoIosArrowBack />
        <p>2022</p>
        <p>Driver Standings</p>
        <div>
          <IoMdMic />
          <AiTwotoneSetting />
        </div>
      </div>
      <input placeholder="Search driver.." onChange={handleChange} />
      <div className="cardsContainer">
        {filteredDrivers.map((driver) => (
          <button className="driverContainer" key={driver.rank} type="button" onClick={() => handleClickItem(`${driver.firstname}${driver.lastname}`)}>
            <TbHelmet />
            <p>{`${driver.firstname} ${driver.lastname}`}</p>
            <p>{`Rank: ${driver.rank}`}</p>
            <p>{`Points: ${driver.points}`}</p>
          </button>
        ))}
      </div>
    </>
  );
};

export default Categories;
