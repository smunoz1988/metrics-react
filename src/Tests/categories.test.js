import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import Categories from '../routes/categories';
import { getCategories } from '../redux/categories/categoriesSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

jest.mock('../redux/categories/categoriesSlice', () => ({
  getCategories: jest.fn(),
}));

describe('Categories Component', () => {
  beforeEach(() => {
    useDispatch.mockClear();
    useSelector.mockClear();
    useNavigate.mockClear();
    getCategories.mockClear();
  });

  test('renders the component with loaded drivers', () => {
    const dispatchMock = jest.fn();
    const navigateMock = jest.fn();
    const drivers = [
      { firstname: 'Lewis', lastname: 'Hamilton', rank: 1, points: 100 },
      { firstname: 'Max', lastname: 'Verstappen', rank: 2, points: 90 },
    ];
    useSelector.mockImplementation((selector) => selector({
      category: {
        categories: {
          drivers: drivers,
        },
      },
    }));
    useDispatch.mockReturnValue(dispatchMock);
    useNavigate.mockReturnValue(navigateMock);

    render(<Categories />);

    expect(screen.getByText('Driver Standings')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Search driver..')).toBeInTheDocument();
    expect(screen.getByText('Lewis Hamilton')).toBeInTheDocument();
    expect(screen.getByText('Max Verstappen')).toBeInTheDocument();

    expect(screen.getByText('Driver Standings')).toMatchSnapshot();
  });

  test('renders the component with loading state', () => {
    const dispatchMock = jest.fn();
    useSelector.mockImplementation((selector) => selector({
      category: {
        loading: true,
        categories: [],
        error: '',
      },
    }));
    useDispatch.mockReturnValue(dispatchMock);

    render(<Categories />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();

    expect(screen.getByText('Loading...')).toMatchSnapshot();
  });
});
