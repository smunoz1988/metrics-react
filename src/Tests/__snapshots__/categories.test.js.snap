// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Component renders the component with loaded drivers 1`] = `
<p>
  Driver Standings
</p>
`;

exports[`Categories Component renders the component with loading state 1`] = `
<div>
  Loading...
</div>
`;
